src/patch_parse.c

```diff
@@ -41,7 +41,19 @@ static int header_path_len(git_patch_parse_ctx *ctx)
 
 	for (len = quoted; len < ctx->parse_ctx.line_len; len++) {
 		if (!quoted && git__isspace(ctx->parse_ctx.line[len]))
-			break;
+			/* lookahead for the sequence "?b/ to allow for spaces in filenames */
+			if (len < ctx->parse_ctx.line_len - 3) {
+				int offset = 0;
+				if (ctx->parse_ctx.line[len+1] == '"') {
+					offset += 1;
+				}
+				if (ctx->parse_ctx.line[len+offset+1] == 'b'
```

Comment:
I'm not that intimate with the diff code, but `b` is not guaranteed to be called `b` (I think there was much sadness about those lines syntax). That's obviously better, but you might want to also not crash on a `diff --git yours/sp ace.txt mine/sp ace.txt` by erroring now (or at end the because there was no `diff --git` line) if we get to `\n` before.

Reply:
Thought as much (I saw some code somewhere where you could name your `a`/`b` anything). Will give it another try.

Reply:
Actually it errors out already if I try `diff --git yours/sp ace.txt mine/sp ace.txt`

```
Started (test status codes: OK='.' FAILURE='F' SKIPPED='S')
.F

  1) Failure:
diff::parse::new_file_with_space_oldnew [/Users/era/git/libgit2/tests/diff/parse.c:128]
  Function call failed: (git_diff_from_buffer(&diff, text, strlen(text)))
  error -1 - corrupt new path in git diff header at line 1
```

Reply:
Are you suggesting that it should error, or be a valid diff header (with `yours/.. theirs/..`)?

Reply:
If it errors cleanly I'm 👍 then.

Reply:
We should be able to parse patches that have custom src and dst path names.  This patch should be valid and parseable:

```
diff --git foo/asdf/new file.txt bar/asdf/new file.txt
new file mode 100644
index 0000000..8bd6648
--- /dev/null
+++ bar/asdf/new file.txt
@@ -0,0 +1 @@
+asdf
```

Reply:
The `diff` header line is a garbage-in-garbage-out horror.  You can set `src-prefix` to `a/b b/` and `dst-prefix` to something like `b/a a b/a/b /b` and it's basically impossible to reason about!

But it appears that git apply:

1. Treats anything up to the first slash as the prefix, and anything after the first slash as the filename.  That is to say that given:

```
diff --git foo/asdf/new file.txt bar/asdf/new file.txt
new file mode 100644
index 0000000..8bd6648
--- /dev/null
+++ bar/asdf/new file.txt
```

it's impossible to tell whether this is a patch of `new file.txt` that was produced with `src-prefix=foo/asdf dst-prefix=bar/asdf`, or a patch of `asdf/new file.txt` that was produced with `src-prefix=foo dst-prefix=bar`.

git is unconcerned with such subtlety and always this as a patch of `asdf/new file.txt` so prefixes must stop at the first slash.

2. Copes fine with prefixes that have a space in them.  eg, the following is legal:

```
diff --git my prefix/new file.txt my prefix/new file.txt
new file mode 100644
index 0000000..8bd6648
--- /dev/null
+++ my prefix/new file.txt
```

Reply:
https://marc.info/?l=git&m=112885453302488&w=2 also describe some kind of quoted C-escape system, which would look like this (mostly, this is an earlier name) : https://marc.info/?l=git&m=112905394808887&w=2